codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/Writable.kt

```diff
@@ -18,14 +18,54 @@ typealias Writable = RustWriter.() -> Unit
  */
 fun writable(w: Writable): Writable = w
 
-fun writable(w: String): Writable = writable { rust(w) }
+fun writable(w: String): Writable = writable { writeInline(w) }
 
 fun Writable.isEmpty(): Boolean {
     val writer = RustWriter.root()
     this(writer)
     return writer.toString() == RustWriter.root().toString()
 }
 
+/**
+ * Helper allowing a `Iterable<Writable>` to be joined together using a `String` separator.
+ */
+fun Iterable<Writable>.join(separator: String) = join(writable(separator))
+
+/**
+ * Helper allowing a `Iterable<Writable>` to be joined together using a `Writable` separator.
+ */
+fun Iterable<Writable>.join(separator: Writable): Writable {
+    val iter = this.iterator()
+    return writable {
+        iter.forEach { value ->
+            value()
+            if (iter.hasNext()) {
+                separator()
+            }
+        }
+    }
+}
```

Comment:
On a related note, I was actually thinking about using the `when` pattern you've used in `rustTypeParameters` rather than using overload resolution to allow for `String` and `Writable` argument.

The reason being, if we wanted to add an `prefix`/`suffix` arguments, in addition to separator, we'd have to write a huge number of different function signatures to allow them to also switch between `Writable` and `String` independently. In contrast, we can just `when` for each argument in one function.

Reply:
I was even thinking of a `ToWritable` interface, which `String`, `RuntimeTypes`, and `Writable` etc satisfies.

Reply:
> I think it could possibly be improved by using this.

https://github.com/awslabs/smithy-rs/compare/harryb/iterable-writable...harryb/shrink-rust-type-params